### DISCUSSION TITLE
Training Loras for Anima

### DISCUSSION THREAD

**Discussion by Fear-in-a-hand-full-of-dust** [67372a/LoRA_Easy_Training_Scripts#69] [Q&A]
Can anyone share some example configs for style lora training on anima? I tried using similar values to my sdxl config, but the results were very underwhelming.

**Comment by uwidev** [comment]
I have been training with OCGOpt, unet only, LR=3e-4. I use gLoRA rank 8 alpha 8, same for convolution layers. I also use `loraplus_lr_ratio=4`. Batch size 4 at 2000 steps with 100 steps of warmup, constant scheduler.

I use a personal script for parsing the config files for batched training, but here's my configs. You probably can't import this directly into the UI, so you'll have to figure out how to change your config to match.
```
[Basics]
pretrained_model_name_or_path = "anima-base-v1.0.safetensors"
qwen3 = "qwen_3_06b_base.safetensors"
vae = "qwen_image_vae.safetensors"
llm_adapter_lr = 0
resolution = "1024"
seed = 23
max_train_steps = 2000

[Save]
output_dir = "output"
output_name = "output"
save_precision = "fp16"
save_model_as = "safetensors"
# save_every_n_epochs = 1
save_every_n_steps = 50
# save_state = true
# save_last_n_steps_state = 1  # basically saving the last + final state if save_state set to true
# save_last_n_epochs_state = 1
# save_n_epoch_ratio = 10
# save_last_n_epochs = 16
# save_last_n_steps = 100000

[SDv2]
scale_v_pred_loss_like_noise_pred = false

[Network_setup]
# lower dim if known style, clean anime style, small dataset, underfitting
# higher dim if unknown style, abstract style, large dataset, overfitting (then lower alpha)
# sane values: 8-24
network_dim = 8
# prefer lower alpha if overfitting, assuming LR is sane
network_alpha = 8
dim_from_weights = false
network_dropout = 0
network_train_unet_only = true
# network_train_text_encoder_only = false
# resume = ''
# network_weights = ''
# base_weights = 'path/to/base_weights'
# base_weights_multiplier = 1

[LyCORIS]
network_module = "lycoris.kohya"
network_args = [

	# conv layers are in charge of local, high frequency content
	# usually should be equal to normal layers, but can half if lines seems to
	# be overfitting (though only if alpha has already been reduced)

	'algo=glora',
	'conv_dim=8',
	'conv_alpha=8',
	'loraplus_lr_ratio=4',
]

[Optimizer]
train_batch_size = 4
gradient_checkpointing = true
gradient_accumulation_steps = 1
optimizer_type = "LoraEasyCustomOptimizer.ocgopt.OCGOpt"
unet_lr = 3e-4
max_grad_norm = 1.0
optimizer_args = []
loss_type = "l2" # default l2
# loss_type = "huber" # default l2

[Lr_scheduler]
lr_scheduler = "constant_with_warmup"
lr_warmup_steps = 100
lr_scheduler_power = 1.0

[Training_preciscion]
mixed_precision = "bf16"
full_bf16 = true

[ARB]
enable_bucket = true
min_bucket_reso = 768
max_bucket_reso = 1536
bucket_reso_steps = 64
bucket_no_upscale = false

[Captions]
shuffle_caption = true
caption_extension = ".txt"
keep_tokens = 0
caption_dropout_rate = 0.05
caption_dropout_every_n_epochs = 0
caption_tag_dropout_rate = 0.0
max_token_length = 150
weighted_captions = false
token_warmup_min = 1
token_warmup_step = 0

[Attention]
sdpa = true

[Data_augmentation]
color_aug = false
flip_aug = false
random_crop = false

[Cache_latents]
cache_latents = true
vae_batch_size = 1
cache_latents_to_disk = true

[Logging]
logging_dir = "logs"
log_with = "tensorboard"
log_prefix = "{datetime}-{name}-"

[Dataset]
max_data_loader_n_workers = 4
persistent_data_loader_workers = true
dataset_repeats = 1                   # Not sure how this is used 

[Regularization]
prior_loss_weight = 1.0

[Debugging]
debug_dataset = false

[Deprecated]
use_8bit_adam = false
use_lion_optimizer = false
learning_rate = 0.0002

[Others]
lowram = false
torch_compile = "True"
disable_cuda_reduced_precision_operations = "True"
loss_related_use_float64 = "True"
training_comment = ""

use_ramtorch="True"
```

```
[[datasets]]
keep_tokens = 0

[[datasets.subsets]]
image_dir = 'input'
num_repeats = 1
shuffle_caption = true
# random_crop = true
# random_crop_padding_percent = 0.05

[general]
resolution = 1024
```

Take what I say with a grain of salt. I'm still very much learning and trying to figure things out, and I don't fully understand the nuances of each parameter and how they affect training.

Results at various training steps.
<img width="3840" height="1457" alt="grid_08" src="https://github.com/user-attachments/assets/706d8d5e-57d2-47ed-8065-64999e63b085" />

**Comment by wongex23** [comment]
> Can anyone share some example configs for style lora training on anima? I tried using similar values to my sdxl config, but the results were very underwhelming.

Would you mind show me your configuration from the GUI?

**Reply by uwidev** [reply]
I haven't actually tested through the UI, but this is the saved toml for what I could populate. Add your dataset and fix the paths to the models and it should work. I would highly suggest you set up some [validation](https://github.com/67372a/LoRA_Easy_Training_Scripts#enhancements-over-original) and logging (I use tensorboard) for your own sanity to ensure training looks good.

```toml
[[subsets]]
caption_extension = ".txt"
name = "subset 1"
num_repeats = 1

[train_mode]
train_mode = "lora"

[general_args.args]
persistent_data_loader_workers = true
pretrained_model_name_or_path = "/mnt/hdd/lib/models/Stable-diffusion/XL/anima-base-v1.0.safetensors"
full_bf16 = true
mixed_precision = "bf16"
gradient_checkpointing = true
seed = 23
max_data_loader_n_workers = 4
max_token_length = 150
prior_loss_weight = 1.0
sdpa = true
max_train_steps = 1000
cache_latents = true
cache_latents_to_disk = true
vae_batch_size = 1
vae = "/mnt/hdd/lib/models/vae/qwen_image_vae.safetensors"

[general_args.dataset_args]
resolution = 1024
batch_size = 4

[network_args.args]
network_dim = 8
network_alpha = 8.0
min_timestep = 0
max_timestep = 1000
network_train_unet_only = true

[textual_inversion_args.args]
token_string = ""
init_word = ""

[optimizer_args.args]
max_grad_norm = 1.0
loss_type = "l2"
optimizer_type = "OCGOptV2"
lr_scheduler = "constant"
learning_rate = 1e-5
warmup_ratio = 0.1

[saving_args.args]
save_precision = "fp16"
save_model_as = "safetensors"

[logging_args.args]
log_prefix_mode = "disabled"
run_name_mode = "default"

[anima_args.args]
qwen3_max_token_length = 512
t5_max_token_length = 512
timestep_sampling = "sigma"
discrete_flow_shift = 3.0
pretrained_model_name_or_path = "/mnt/hdd/lib/models/stable-diffusion/XL/anima-base-v1.0.safetensors"
qwen3 = "/mnt/hdd/lib/models/text_encoders/qwen_3_06b_base.safetensors"
vae = "/mnt/hdd/lib/models/vae/qwen_image_vae.safetensors"

[edm_loss_args.args]
edm2_loss_weighting = false

[bucket_args.dataset_args]
enable_bucket = true
bucket_reso_steps = 64
min_bucket_reso = 768
max_bucket_reso = 1536
bucket_no_upscale = true

[network_args.args.network_args]
conv_dim = 8
conv_alpha = 8.0
algo = "glora"
loraplus_lr_ratio = "4"
use_scalar = "True"

[optimizer_args.args.optimizer_args]
foreach = "True"
```

**Reply by uwidev** [reply]
Also consider seeing https://github.com/67372a/LoRA_Easy_Training_Scripts/discussions/84

**Reply by wongex23** [reply]
Thanks for your sample.

I tried below also looks good. 

But that will cost me about 4 hours for just1 lora. 

I am using 3060 12Gb Vram. Nou sure if anything we can make it faster.


[[subsets]]
caption_dropout_rate = 0.1
caption_extension = ".txt"
caption_tag_dropout_rate = 0.15
image_dir = "E:/AI/AI_SD/PicForTrain/Maid"
name = "subset 1"
num_repeats = 4
random_crop_padding_percent = 0.0
shuffle_caption = true
token_warmup_min = 1
token_warmup_step = 1

[train_mode]
train_mode = "lora"

[general_args.args]
persistent_data_loader_workers = true
vae_batch_size = 1
pretrained_model_name_or_path = ""
lowram = true
full_bf16 = true
mixed_precision = "bf16"
gradient_checkpointing = true
gradient_accumulation_steps = 1
seed = 168
max_data_loader_n_workers = 1
max_token_length = 225
prior_loss_weight = 1.0
sdpa = true
cache_latents = true
keep_tokens_separator = "|||"
max_train_steps = 2000

[general_args.dataset_args]
resolution = 896
batch_size = 2

[network_args.args]
network_dim = 16
network_alpha = 8.0
min_timestep = 0
max_timestep = 1000

[optimizer_args.args]
optimizer_type = "Came"
lr_scheduler = "cosine"
learning_rate = 5e-5
warmup_ratio = 0.05
unet_lr = 5e-5
text_encoder_lr = 6e-6
max_grad_norm = 0.8
min_snr_gamma = 8.0
loss_type = "l2"

[saving_args.args]
output_dir = "E:/AI/AI_SD/Trained/Lora"
output_name = "Maid_CAME_ANIMAv3"
save_precision = "bf16"
save_model_as = "safetensors"
save_every_n_steps = 100

[sample_args.args]
sample_sampler = "euler_a"
sample_prompts = "E:/AI/AI_SD/SampleText/Maid.txt"
sample_every_n_steps = 100

[logging_args.args]
log_prefix_mode = "disabled"
run_name_mode = "default"

[anima_args.args]
qwen3 = "E:/AI/ForgeUI_OLD/webui/models/text_encoder/qwen_3_06b_base.safetensors"
vae = "E:/AI/ForgeUI_OLD/webui/models/VAE/qwen_image_vae.safetensors"
qwen3_max_token_length = 512
t5_max_token_length = 512
timestep_sampling = "sigmoid"
sigmoid_scale = 1.0
attn_mode = "flash"
split_attn = true
pretrained_model_name_or_path = "E:/AI/ForgeUI_OLD/webui/models/Stable-diffusion/Anima_Backup/anima_preview3Base.safetensors"

[edm_loss_args.args]
edm2_loss_weighting = false

[bucket_args.dataset_args]
enable_bucket = true
multires_training = true
min_bucket_reso = 256
max_bucket_reso = 1536
bucket_reso_steps = 64

[network_args.args.network_args]
conv_dim = 16
conv_alpha = 8.0
algo = "locon"

[optimizer_args.args.optimizer_args]
weight_decay = "0.01"

**Reply by uwidev** [reply]
Here are some other things you can do for faster training (for context, I train styles, may not apply fully to other concepts).
- If you're finetuning an existing tag, caption all images with that tag and use a unique unknown tag. This tag will strictly take on the deltas rather than trying to fight against the weights of the existing tag. This will speed up learning.
- If the tag is already known, you should be good results within about 1000 steps. 2000 is overkill and only if you're looking for the finest of details. If you really want, you can go for 1500 for padding, or if your're confident, 700.
- `lora_plus` and `use_scalar` can make training faster as are further multipliers on top of the LoRA. Can lead to faster convergence but requires but requires some tuning on the learning rate.
- I train only UNET only, very rarely do I touch TE. You can train faster with UNET only.
- Seriously try OCGOptv2. It makes early training steps still usable. I remember Prodigy and CAME days early training steps ~500 were essentially unusable. But OCGOpt(and v2) have safeguards to have outputs from training steps ~200 be usable and coherent, even at large learning rates (up to a limit). It can help you better search for a proper early stopping point for your needs. This DOES require better batching, I recommend 4 total, but 2 is workable.
- Your minimum bucket resolution is too low. Unless you did it intentionally, the lowest should be at 768, roughly x0.75 down from 1024. This should reduce the total bucketing; less buckets = more images per buckets = faster training but slightly less generalization. It should be fine as for Anima you're usually not generating anything below 512x512.
- Also, since you're VRAM limited, use [RamTorch](https://github.com/67372a/LoRA_Easy_Training_Scripts#11172025) to offload to system RAM. Can be slightly slower, but at least you can do higher batches. For OCGOpt, it's crucial. You can also play games and do stuff on GPU since you have more VRAM now.
